src/app-config/auth0_shim.ts

```diff
@@ -0,0 +1,318 @@
+import atob from "atob";
+import base64url from "base64url";
+import btoa from "btoa";
+import * as Crypto from 'crypto';
+
+export interface Auth0Transaction {
+  scope: string;
+  audience: string;
+  code_verifier: string;
+  authorizeOptions: any;
+  state: string;
+  nonce: string;
+  code_challenge: string;
+  redirect_uri?: string;
+  url: string;
+}
+
+export interface RedirectLoginOptions {
+  redirect_uri?: string;
+  scope?: string;
+  fragment?: string;
+}
+
+/**
+ * This auth0-shim is used to support building the auth0 "/authorize" url. This is typically done in the browser
+ * and auth0 packaged that code in the auth0-spa-js library: https://github.com/auth0/auth0-spa-js
+ *
+ * Unfortunately, the auth0-spa-js library is not written with "Universal javascript" in mind and leans on the dom for certain
+ * string manipulation methods (ie: https://github.com/auth0/auth0-spa-js/blob/de3f3d0b400875883e161bbcda7094faf1e50933/src/utils.ts#L212)
+ *
+ * looking across the auth0 ecosystem, I don't see support for the browswer login flow that doesn't make assumptions that
+ * the url is being generated in a browser, so I've rolled our own here (largely taking code snippets from auth0-spa-js)
+ *
+ * Opened an issue with them here: https://github.com/auth0/auth0-spa-js/issues/432
+ */
+export class Auth0Shim {
+
+  public static build_auth0_transaction(
+    auth0_client_id: string,
+    domain: string,
+    options: RedirectLoginOptions,
+  ): Auth0Transaction {
+    const { redirect_uri, ...authorizeOptions } = options;
+
+    const stateIn = btoa(Auth0Shim.createRandomString());
+    const nonceIn = btoa(Auth0Shim.createRandomString());
+    const code_verifier = Auth0Shim.createRandomString();
+
+    const code_challengeBuffer = Crypto.createHash('sha256').update(code_verifier).digest();
+    const code_challenge = base64url(code_challengeBuffer);
+    const fragment = options.fragment ? `#${options.fragment}` : '';
+
+    const params = {
+      ...authorizeOptions,
+      client_id: auth0_client_id,
+      scope: this.getUniqueScopes(
+        'openid profile email',
+        authorizeOptions.scope ? authorizeOptions.scope : ''
+      ),
+      response_type: 'code',
+      response_mode: 'query',
+      audience: 'architect-hub-api', //TODO:46:
```

Comment:
@tjhiggins do you foresee us needing api access from the CLI? If not I think I can remove this....